src/keychain.js

```diff
@@ -400,99 +369,96 @@ class Keychain {
-            batch.put(dsname, pem)
-            batch.put(DsInfoName(name), JSON.stringify(keyInfo))
-            batch.commit((err) => {
-              if (err) return _error(callback, err)
-
-              callback(null, keyInfo)
-            })
-          })
-        })
-      })
-    })
+    const exists = await self.store.has(dsname)
+    if (exists) return throwDelayed(errcode(new Error(`Key '${name}' already exists`), 'ERR_KEY_ALREADY_EXISTS'))
+
+    let privateKey
+    try {
+      privateKey = await crypto.keys.import(pem, password)
+    } catch (err) {
+      return throwDelayed(errcode(new Error('Cannot read the key, most likely the password is wrong'), 'ERR_CANNOT_READ_KEY'))
+    }
+
+    let kid
+    try {
+      kid = await privateKey.id()
+      pem = await privateKey.export(this._())
+    } catch (err) {
+      return throwDelayed(err)
+    }
+
+    const keyInfo = {
+      name: name,
+      id: kid
+    }
+    const batch = self.store.batch()
+    batch.put(dsname, pem)
+    batch.put(DsInfoName(name), JSON.stringify(keyInfo))
+    await batch.commit()
+
+    return keyInfo
   }
 
-  importPeer (name, peer, callback) {
+  async importPeer (name, peer) {
     const self = this
     if (!validateKeyName(name)) {
-      return _error(callback, errcode(new Error(`Invalid key name '${name}'`), 'ERR_INVALID_KEY_NAME'))
+      return throwDelayed(errcode(new Error(`Invalid key name '${name}'`), 'ERR_INVALID_KEY_NAME'))
     }
     if (!peer || !peer.privKey) {
-      return _error(callback, errcode(new Error('Peer.privKey is required'), 'ERR_MISSING_PRIVATE_KEY'))
+      return throwDelayed(errcode(new Error('Peer.privKey is required'), 'ERR_MISSING_PRIVATE_KEY'))
     }
 
     const privateKey = peer.privKey
     const dsname = DsName(name)
-    self.store.has(dsname, (err, exists) => {
-      if (err) return _error(callback, err)
-      if (exists) return _error(callback, errcode(new Error(`Key '${name}' already exists`), 'ERR_KEY_ALREADY_EXISTS'))
-
-      privateKey.id((err, kid) => {
-        if (err) return _error(callback, err)
-        privateKey.export(this._(), (err, pem) => {
-          if (err) return _error(callback, err)
-          const keyInfo = {
-            name: name,
-            id: kid
-          }
-          const batch = self.store.batch()
-          batch.put(dsname, pem)
-          batch.put(DsInfoName(name), JSON.stringify(keyInfo))
-          batch.commit((err) => {
-            if (err) return _error(callback, err)
-
-            callback(null, keyInfo)
-          })
-        })
-      })
-    })
+    const exists = await self.store.has(dsname)
+    // if (err) return throwDelayed(callback, err)
+    if (exists) return throwDelayed(errcode(new Error(`Key '${name}' already exists`), 'ERR_KEY_ALREADY_EXISTS'))
+
+    const kid = await privateKey.id()
+    // if (err) return throwDelayed(callback, err)
+    const pem = await privateKey.export(this._())
+    // if (err) return throwDelayed(callback, err)
```

Comment:
Should we remove all the above comments?

We should probably add a `try...catch` from the `const kid` line until the commit throwing using the throwDelayed if we catch any error?

Reply:
Done, I also wrapped renameKey to make sure we're delaying there as well.